### DISCUSSION TITLE
jvm/neicun-jiegou

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#223] [Announcements]
# jvm/neicun-jiegou

Java程序员进阶之路，小白的零基础Java教程，从入门到进阶，深入理解JVM的内存结构

https://tobebetterjavaer.com/jvm/neicun-jiegou.html

**Comment by light-dot** [comment]
“第一，永久代放在 Java 虚拟机中，就会受到 Java 虚拟机内存大小的限制，而元空间使用的是本地内存，也就脱离了 Java 虚拟机内存的限制。”关于这一条，如果说元空间不再JVM的内存中，那是不是就是在JVM所运行的这个主机上的内存，如果这个主机上运行了很多java程序，这些java程序的元空间都没有约束，是不是会发生资源抢占的情况，从而导致程序异常？

**Reply by chaofengdev** [reply]
真是个好问题。不过我也不知道答案。

**Reply by chaofengdev** [reply]
为了防止元空间占用过多的本地内存，可以通过以下 JVM 参数进行控制：

-XX:MaxMetaspaceSize=<size>：限制元空间的最大大小。如果不指定该参数，元空间默认是动态扩展的，直到达到系统可用内存的上限。
-XX:MetaspaceSize=<size>：设定元空间的初始大小。JVM 在启动时会根据该参数来为元空间预留内存，当达到这个大小时，JVM 会进行垃圾回收（GC），清理无用的类元数据。
通过合理设置这些参数，可以有效避免多个程序争夺主机内存资源，从而减少资源抢占带来的问题。
